earthmover/earthmover.py

```diff
@@ -383,3 +391,155 @@ def test(self, tests_dir: str):
             if not _expected_df.equals(_outputted_df):
                 self.logger.critical(f"Test output `{_outputted_file}` does not match expected output.")
                 exit(1)
+
+
+    def deps(self):
+        """
+        Installs all packages specified in the config file and any nested packages.
+        :return:
+        """
+        self.build_root_package_graph()
+
+        # Check that at least one package is defined
+        if all(False for _ in self.package_graph.successors('root')):
+            self.error_handler.throw("No packages have been defined!")
+
+        # Install each package (and any nested sub-packages) into the packages directory
+        self.build_package_graph(root_node='root', package_subgraph=self.package_graph, packages_dir=self.packages_dir, install=True)
+
+
+    def merge_packages(self):
+        """
+        Traverses the packages graph, merging yaml config from successors into predecessors.
+        Saves the final result as the instance user_configs.
+        :return:
+        """
+        self.build_root_package_graph()
+
+        # If the yaml file doesn't include packages, no need to alter
+        if all(False for _ in self.package_graph.successors('root')):
+            return
+        
+        self.build_package_graph(root_node='root', package_subgraph=self.package_graph, packages_dir=self.packages_dir, install=False)
+
+        # Merge each package yaml into the predecessor yaml, storing the result in the predecessor
+        # Post-order traversal ensures the correct hierarchy of merges
+        for package_name in nx.dfs_postorder_nodes(self.package_graph):
+            node_package = self.package_graph.nodes[package_name]['package']
+
+            for predecessor_name in self.package_graph.predecessors(package_name): # more elegant way to do this? we know each node will only have one predecessor
+                predecessor_package = self.package_graph.nodes[predecessor_name]['package']
+                
+                # Load package yaml if not yet loaded
+                node_yaml = node_package.package_yaml or node_package.load_package_yaml(self.params, self.macros)
+                predecessor_yaml = predecessor_package.package_yaml or predecessor_package.load_package_yaml(self.params, self.macros)
+
+                merged_yaml = self.merge_dicts(node_yaml, predecessor_yaml)
+                predecessor_package.package_yaml = merged_yaml
+
+
+        # Overwrite with completed merged yaml  
+        self.user_configs = self.package_graph.nodes['root']['package'].package_yaml
+        
+        # Output merged yaml file

```

Comment:
Should this write occur here? When there are no packages, the parsed YAML file is not written. I'd argue that a write should always occur to visualize Jinja templating.